client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+
+* **Question(5)**: How do we format batch writes as an HTTP request?
+
+### Responses
+As with the Requests section, please see more information in [OMVCC and Caching](https://forum.treode.com/t/omvcc-and-http-caching/62).
+
+#### Read Respones
+
+The DB will respond to read requests with a message containing the desired value or an indication the value does not exist in the table. 
+
+In the success case, HTTP response for read should be similar to the following:
+
+    HTTP/1.1 200 Okay
+    Date: Wed, 14 Jan 2015 11:49:13 GMT
+    Last-Modified: Fri, 10 May 2013 02:07:43 GMT
+    Read-TxClock: 1421236153024853
+    Value-TxClock: 1368151663681367
+    Vary: Request-TxClock
+    {   "title": "Star Wars",
+        "cast": [
+            { "actor": "Mark Hamill", "role": "Luke Skywalker" },
+            { "actor": "Carrie Fisher", "role": "Princess Leia Organa" } ] }
+
+There are several cases in which the HTTP response does not include the JSON-formatted requested value.  Here are those cases:
+
+- **HTTP/1.1 304 Not Modified**: The requested value has not changed since the last request.  Note this response is still valuable to the cache, as it allows the cache to update the read time (R_t) for this entry to the read\_timestamp of this request.  (Between when the value was written and this read\_timestamp, we know the value has not changed.)
+
+- **HTTP/1.1 404 Not Found**: The requested value is not in the table.  This response is also valuable to the cache, as the cache can store None (in Python, possibly null in other languages) in the JSON entry for this entry at the given read\_timestamp.  Then future requests see the key had no value at this time without going to the DB server.
+
+#### Write Responses
+
+The DB will respond with either a success or rejection indicator.  If the DB rejects, the cache write method should throw an Stale Exception, indicating to the caller to retry the transaction with more up-to-date values from the DB.
+
+The HTTP request for write should be similar to the following:
+
+    HTTP/1.1 200 Okay
+    Value-TxClock: 1421236393816753
+
+However, it is also possible the DB server will be unable to complete the write because another client has written new values in the meantime.  In this case, the HTTP write resposne will be of the following form:
+
+    HTTP/1.1 412 Precondition Failed 
+
+In this case, the the client should retry the transaction with a smaller value for the min-age header (e.g., min-age=0) to ensure the values it receives from the DB are not stale.  
+
+<a name="Transaction"></a>
+# Transaction
+
+- Parameters: 
+    1. cache: Cache (required)
+    2. read\_timestamp: int (read\_timestamp=None)
+    3. max\_age: int (max\_age=None)
+    4. no\_cache: bool (no\_cache=False)
+- Public Methods: 
+    1. read(table: string, key: string, read\_timestamp=currentTime, max\_age=None, no\_cache=False) returns value (JSON) or None
```

Comment:
Tx.read always uses the read time from the Tx constructor. There's no reason to let the user override it read by read. That just creates confusion.